docs: more content around transactions, retries and merge behaviour (#1065)
neo4j-contrib/neo4j-spark-connector: docs/modules/ROOT/pages/write/transactions.adoc

= Transactions, retries, and partial failures

:description: How the connector splits writes into transactions, what happens when a write fails halfway through, and how to write safely with MERGE under parallelism.

include::partial$sparksession.adoc[]

A write from Spark to Neo4j is not a single transaction.
Understanding how the connector splits a write into many transactions explains what you are left with when a write job fails halfway through, and why `MERGE` needs extra care when more than one partition is writing at the same time.

[#transaction-model]

> == How a write is split into transactions

Reviewer: ```suggestion
== Transaction model
```

For every write, the connector builds one Cypher query and executes it once per batch of rows:

> * Spark runs one *task* per DataFrame partition, and each task opens its own session to Neo4j.

Reviewer: ```suggestion
* Spark runs one task per DataFrame partition, and each task opens its own session to Neo4j.
```

* Each task accumulates rows until it has `batch.size` rows, then opens a transaction, runs the query with that batch bound to the `$events` parameter, and commits.
* The task repeats this until the partition is exhausted. A final, possibly smaller, batch is committed when the task completes.

So a job writing 1 million rows with `batch.size` set to 5000 and four partitions commits roughly 200 independent transactions, up to four of them concurrently.

[IMPORTANT]
====
There is no enclosing transaction around a write, and no coordination between partitions.
Each batch is committed on its own and becomes visible to other clients immediately.
====

This has a direct consequence: *the connector cannot offer all-or-nothing writes*. Sizing `batch.size` is therefore both a performance decision (see xref:performance/tuning.adoc#batch-size[]) and a decision about how much work can be lost or duplicated if a write fails.

[#call-in-transactions]

> == `CALL { … } IN TRANSACTIONS` is not supported

Reviewer: ```suggestion
== `CALL` subqueries in transactions
```

Use `batch.size` instead. It is the connector's equivalent of `IN TRANSACTIONS OF N ROWS`, and it applies to every write option, not just `query`.

[#partial-failures]
== Partial failures

When a write job fails, the data already in the database stays there. Specifically:

* Batches that were committed before the failure are *not* rolled back.
* The batch in flight when the error was raised is rolled back, so it applies fully or not at all.
* Other partitions are unaffected and keep writing until they finish or fail on their own.

The result is a database holding an arbitrary prefix of each partition's rows, with no marker recording where each one stopped.
Rerunning the job replays the rows that did succeed.

[WARNING]
====
Never treat a failed write as a no-op.
Before rerunning a failed job, either make the write idempotent (see <<merge-safety>>) or clean up what the previous attempt committed, otherwise the retry duplicates the rows that had already landed.
====

=== Spark task retries

Spark's own fault tolerance assumes tasks can be replayed safely, which does not hold for a non-transactional sink.

> If a task throws, Spark may retry the whole task, which re-runs the *entire partition* from its first row, including every batch the failed attempt had already committed.

Reviewer: ```suggestion
If a task fails with an exception, Spark may retry the whole task, which re-runs the *entire partition* from its first row, including every batch the failed attempt had already committed.
```

With `SaveMode.Append`, which generates `CREATE`, every replayed row becomes a duplicate.

To make a task retry harmless, use `SaveMode.Overwrite` with keys and a supporting constraint, so replaying a row updates it instead of creating a second copy.

=== Streaming writes

Structured Streaming writes use the same per-batch commit model and the connector's epoch commit is a no-op, so streaming gives you the same at-least-once behavior described above. A failed and restarted epoch can rewrite rows it had already committed.

[#retries]
== Retry behavior

Each task retries a failed batch itself before giving up, controlled by three options:

.Retry options
[cols="1,3,1"]
|===
|Option |Description |Default

|`transaction.retries`
|How many times the task may retry a failed batch.
|`3`

|`transaction.retry.timeout`
|Milliseconds to wait before each retry.
|`0`

|===

The logic applied to an exception is, in order:

> . If the exception is retryable and the task's retry budget is not exhausted, the task closes its transaction and session, waits `transaction.retry.timeout`, and re-runs *the current batch only*.

Reviewer 2: meta question: do we really need to close the session here? isn't closing the transaction enough?
we'd lose the accumulated bookmarks for that partition's session by creating a new session, or?

Reviewer 2: causality is probably not a big deal since work is not coordinated across partitions, I guess it'd still be nice to maintain causality within partitions however

Author: Does causality really matter within a read only or write only context? I think it's more important when you have mixed read and write workload, which we don't?

If a connection is lost while a commit is in flight, the outcome is *indeterminate*: the server may have committed the transaction and been unable to tell the client.
The driver raises `ServiceUnavailableException`, the connector classifies it as retryable, and the batch is written again.
This is the mechanism behind duplicates that appear after `ExclusiveLock` and connection errors in the logs, even though the job reports the batch as failed and retried.

Which outcomes are possible depends on how the write is expressed:

[cols="1,2"]
|===
|Write |Effect of a retry after an indeterminate commit

|`SaveMode.Append` (`CREATE`)
|Duplicate nodes or relationships, always. `CREATE` has no notion of an existing row.

|`SaveMode.Overwrite` (`MERGE`) with a uniqueness or key constraint on the keys
|Safe. The retry matches what the first attempt wrote and updates it.

|`SaveMode.Overwrite` (`MERGE`) without a constraint on the keys
|Duplicates are possible, because `MERGE` may not see a concurrently written node. See <<merge-safety>>.

|`query` option
|Depends entirely on your query. Only an idempotent query survives a retry.
|===

[#merge-safety]
== MERGE safety under parallelism

> `MERGE` looks idempotent, and is, in a single-threaded write against a constrained property.

Reviewer 2: ```suggestion
`MERGE` looks idempotent, and is, in a single-threaded single-writer write against a constrained property.
```
this is a nit, but if you've got different single-threaded processes running `MERGE`, that does not hold anymore

  .save()
----

* *Write nodes before relationships*, in separate jobs. Node writes partition cleanly by key; relationship writes lock both endpoints and are far harder to partition safely.

* *Use a single partition for relationship writes*, unless you can guarantee that no two partitions share an endpoint node. See xref:performance/tuning.adoc#parallelism[].

[#large-imports]
== Recommended pattern for large imports

Putting the above together, a large import that tolerates failure and retry looks like this:

. *Create constraints first*, in their own step, using the schema optimization options or a `script` option. This is what makes each row's write idempotent.
. *Write nodes, then relationships*, as separate jobs, so relationship writes can always find their endpoints and node writes stay partitionable.
. *Repartition by key* for node writes; use one partition for relationship writes unless the data is partitioned to avoid shared endpoints.
. *Use `SaveMode.Overwrite` with keys* rather than `Append`, so a replayed batch or task updates instead of duplicating.
. *Size `batch.size` to the server's heap*, and remember it also bounds how much work a single failure can lose.

> . *Raise `transaction.retries` and set a non-zero `transaction.retry.timeout`* for long partitions, so a leader switch or a burst of lock contention does not fail the job.

Reviewer: I'd skip highlighting the start of each sentence in bold here.

  // the relationship instead of creating a second one
  .option("relationship.keys", "since")
  .option("transaction.retries", "10")
  .save()
----

[#memory]
== Keep per-row work bounded

A batch is one transaction, and its whole working set must fit in the server's heap.
Queries that expand the amount of work per row can make a nominally reasonable `batch.size` unaffordable, which shows up as out-of-memory errors or, on Aura, a quarantined instance rather than a clean failure.

Patterns that expand per-row work include:

* Setting an entire map of unknown size, as in `SET n += event.properties`. The transaction has to hold every property of every row in the batch, and the property count is data-dependent rather than schema-bounded.
* Merging or matching on unconstrained or unindexed properties, which turns each row into a scan.
* Creating relationships whose endpoints are matched by a non-key property.

If a write runs out of memory, lower `batch.size` first, then constrain what each row does: project only the columns you need before writing, and prefer explicit properties over whole-map assignment so that the per-row cost is bounded by your schema.

> == See also

Reviewer: ```suggestion
== Resources
```

Reviewer: Besides renaming, is this section useful or can these links be embedded better within the other sections?

Author: Removed this section.

= Transactions, retries, and partial failures

:description: How the connector splits writes into transactions, what happens when a write fails halfway through, and how to write safely with MERGE under parallelism.

include::partial$sparksession.adoc[]

A write from Spark to Neo4j is not a single transaction.
Understanding how the connector splits a write into many transactions explains what you are left with when a write job fails halfway through, and why `MERGE` needs extra care when more than one partition is writing at the same time.

[#transaction-model]
== How a write is split into transactions

For every write, the connector builds one Cypher query and executes it once per batch of rows:

* Spark runs one *task* per DataFrame partition, and each task opens its own session to Neo4j.

> * Each task accumulates rows until it has `batch.size` rows (5000 by default), then opens a transaction, runs the query with that batch bound to the `$events` parameter, and commits.

Reviewer 3: ```suggestion
* Each task accumulates rows until it has `batch.size` rows, then opens a transaction, runs the query with that batch bound to the `$events` parameter, and commits.
```

I think we risk having parity between actual and documented if we put default value here. I'm sure we'll forget to change on two places should we ever change the batch size.

* The task repeats this until the partition is exhausted. A final, possibly smaller, batch is committed when the task completes.

> So a job writing 1 million rows with the default `batch.size` and four partitions commits roughly 200 independent transactions, up to four of them concurrently.

Reviewer 3: ```suggestion
So a job writing 1 million rows with `batch.size` set to 5000 and four partitions commits roughly 200 independent transactions, up to four of them concurrently.
```

If we remove above, we should be explicit here. Also if we change default batch size, this paragraph changes

[#call-in-transactions]
== `CALL { … } IN TRANSACTIONS` is not supported

Because the connector runs your query inside an explicit transaction, you cannot use Cypher's link:{neo4j-docs-base-uri}/cypher-manual/current/subqueries/subqueries-in-transactions/[`CALL { … } IN TRANSACTIONS`^] to do your own batching, in either the `query` option or the `script` options.
Neo4j rejects that clause inside an explicit transaction, so the write fails with an error.

Two further reasons it cannot work with the `query` option:

* Your query is not executed standalone. It is appended to an `UNWIND $events AS event` clause, so it is a query fragment and not a place where a top-level `CALL { … } IN TRANSACTIONS` clause could appear.
* Your query never sees the whole DataFrame, only the current batch, so there would be nothing left for it to subdivide.

Use `batch.size` instead. It is the connector's equivalent of `IN TRANSACTIONS OF N ROWS`, and it applies to every write option, not just `query`.

[#partial-failures]
== Partial failures

When a write job fails, the data already in the database stays there. Specifically:

> * Batches that were committed before the failure are *not* rolled back. There is nothing to roll them back with.

Reviewer 3: ```suggestion
* Batches that were committed before the failure are *not* rolled back.
```

Think we can remove this and get the same message across

Before rerunning a failed job, either make the write idempotent (see <<merge-safety>>) or clean up what the previous attempt committed, otherwise the retry duplicates the rows that had already landed.
====

=== Spark task retries

Spark's own fault tolerance assumes tasks can be replayed safely, which does not hold for a non-transactional sink.
If a task throws, Spark may retry the whole task, which re-runs the *entire partition* from its first row, including every batch the failed attempt had already committed.
With `SaveMode.Append`, which generates `CREATE`, every replayed row becomes a duplicate.

To make a task retry harmless, use `SaveMode.Overwrite` with keys and a supporting constraint, so replaying a row updates it instead of creating a second copy.

=== Streaming writes

Structured Streaming writes use the same per-batch commit model and the connector's epoch commit is a no-op, so streaming gives you the same at-least-once behavior described above. A failed and restarted epoch can rewrite rows it had already committed.

[#retries]
== Retry behavior

Each task retries a failed batch itself before giving up, controlled by three options:

> .Retry options

Reviewer 3: oh boy okay we need to remember to parity-update this page in case we ever change defaults... unless we can get away with removing defaults here as well, or somehow use anotra/adoc partials to define this section of the table in one place.

Author: If the tables were in the same shape, I think partials could help but it looks like they are slightly different in each page. So, I think I'll keep it as it is.
